cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java

```diff
@@ -149,7 +159,62 @@ private Plugin findPlugin(String pluginType, String pluginName, String pluginId,
     Preconditions.checkArgument(!plugins.containsKey(pluginId),
                                 "Plugin of type %s, name %s was already added as id %s.",
                                 pluginType, pluginName, pluginId);
-    return FindPluginHelper.findPlugin(artifactRepository, pluginInstantiator, deployNamespace.toEntityId(), artifactId,
-                                       pluginType, pluginName, properties, selector);
+
+
+    final Class[] callerClasses = CallerClassSecurityManager.getCallerClasses();
+    if (callerClasses.length < 3) {
+      // This shouldn't happen as there should be someone calling this method.
+      throw new IllegalStateException("Invalid call stack.");
+    }
+
+    Set<ArtifactId> seenParents = new HashSet<>();
+    List<ArtifactId> parents = new ArrayList<>();
+    // go through the call stack to get the parent plugins. This is to support plugins of plugins.
+    // for example, suppose appX uses pluginY, which uses pluginZ. When we try to use pluginZ, we need to first
+    // check if pluginZ extends pluginY. If it does not, we then check if it extends appX.
+    // 0 is the CallerClassSecurityManager, 1 is this class, hence 2 is the actual caller
+    for (int i = 2; i < callerClasses.length; i++) {
+      ClassLoader callerClassloader = callerClasses[i].getClassLoader();
+      if (AbstractPluginConfigurable.class.isAssignableFrom(callerClasses[i]) || callerClassloader == null) {
+        // we've reached the program, no need to go higher up the call stack.
+        break;
+      } else if (callerClassloader instanceof PluginClassLoader) {
+        // if this is the first time we've seen this plugin artifact, it must be a new plugin parent.
+        ArtifactId pluginCallerArtifactId = ((PluginClassLoader) callerClassloader).getArtifactId();
+        if (seenParents.add(pluginCallerArtifactId)) {
+          parents.add(pluginCallerArtifactId);
```

Comment:
had the list since that's what the Plugin class uses for its parents object. But with your other comment, I'll use a LinkedHashSet for seenParents and then add the app artifact id to it to just have a single loop below.